Closes #2441 refactor for modular design

imccay13 opened the pull request:
In response to https://github.com/pharmaverse/admiral/discussions/2349, refactors derive_vars_merged


Thank you for your Pull Request! We have developed this task checklist from the [Development Process Guide](https://pharmaverse.github.io/admiral/CONTRIBUTING.html#detailed-development-process) to help with the final steps of the process. Completing the below tasks helps to ensure our reviewers can maximize their time on your code as well as making sure the admiral codebase remains robust and consistent.   

Please check off each taskbox as an acknowledgment that you completed the task or check off that it is not relevant to your Pull Request. This checklist is part of the Github Action workflows and the Pull Request will not be merged into the `main` branch until you have checked off each task.

- [x] Place Closes #<insert_issue_number> into the beginning of your Pull Request Title (Use Edit button in top-right if you need to update)
- [x] Code is formatted according to the [tidyverse style guide](https://style.tidyverse.org/). Run `styler::style_file()` to style R and Rmd files
- [x] Updated relevant unit tests or have written new unit tests, which should consider realistic data scenarios and edge cases, e.g. empty datasets, errors, boundary cases etc. - See [Unit Test Guide](https://pharmaverse.github.io/admiraldev/articles/unit_test_guidance.html#tests-should-be-robust-to-cover-realistic-data-scenarios)
- [x] If you removed/replaced any function and/or function parameters, did you fully follow the [deprecation guidance](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#deprecation)?
- [x] Review the [Cheat Sheet](https://github.com/pharmaverse/admiral/blob/main/inst/cheatsheet/admiral_cheatsheet.pdf). Make any required updates to it by editing the file `inst/cheatsheet/admiral_cheatsheet.pptx` and re-upload a PDF version of it to the same folder.
- [x] Update to all relevant roxygen headers and examples, including keywords and families. Refer to the [categorization of functions](https://pharmaverse.github.io/admiraldev/articles/programming_strategy.html#categorization-of-functions) to tag appropriate keyword/family.
- [x] Run `devtools::document()` so all `.Rd` files in the `man` folder and the `NAMESPACE` file in the project root are updated appropriately
- [x] Address any updates needed for vignettes and/or templates
- [x] Update `NEWS.md` under the header `# admiral (development version)` if the changes pertain to a user-facing function (i.e. it has an `@export` tag) or documentation aimed at users (rather than developers). A Developer Notes section is available in `NEWS.md` for tracking developer-facing issues.  
- [x] Build admiral site `pkgdown::build_site()` and check that all affected examples are displayed correctly and that all new functions occur on the "[Reference](https://pharmaverse.github.io/admiral/reference/index.html)" page. 
- [x] Address or fix all lintr warnings and errors - `lintr::lint_package()`
- [x] Run `R CMD check` locally and address all errors and warnings - `devtools::check()`
- [x] Link the issue in the Development Section on the right hand side.
- [x] Address all merge conflicts and resolve appropriately
- [x] Pat yourself on the back for a job well done! Much love to your accomplishment!

bms63 commented:
Hi @imccay13 I see this is draft, but just wanted to say there seems to be quite a few files that are not needed for this update in this PR.

imccay13 commented:
> Hi @imccay13 I see this is draft, but just wanted to say there seems to be quite a few files that are not needed for this update in this PR.

@bms63 Believe I fixed this, and also fixed the problem I was having with the vignettes - let me know if I'm right about those two things and I'll open the PR

bms63 commented:
Hi @imccay13 if they pass all the checks (you can ignore link one), then please click ready for review.  For the new function, I'm just glancing at this PR so far, I see only one unit test for it.  Do you think that is sufficient coverage for the new function?  I'd might error on the side of caution and have some dirty data scenarios and checks for messaging to users??

@bundfussr and @manciniedoardo can we prioritize a review as it might impact on admiralroche.

manciniedoardo commented:
@millerg23 @bundfussr please could you give this a look-over with admiralroche in mind? thanks

bundfussr commented on R/derive_vars_query.R:
```diff
@@ -58,8 +61,8 @@
-#' derive_vars_query(adae, queries)
-derive_vars_query <- function(dataset, dataset_queries) { # nolint: cyclocomp_linter
+#' derive_vars_for_query(adae, queries)
+derive_vars_for_query <- function(dataset, dataset_queries) { # nolint: cyclocomp_linter
```
The function needs to be renamed because it removes variables from the input dataset. I would suggest naming it `get_vars_query()`.

bundfussr commented on R/derive_vars_query.R:
```diff
@@ -1,5 +1,8 @@
 #'
+#' **Note:** This function exists to work with derive_vars_query, to execute
+#' the first step of derive_vars_query separately for metadata tracking
+#'
```
The `@return` section needs to be updated because not all variables of the input dataset are included in the output dataset.

`@family` and `@keywords` should be changed to `utils_help` because the function is not a derivation.

bundfussr commented on R/derive_vars_query.R:
```diff
@@ -193,7 +196,86 @@ derive_vars_query <- function(dataset, dataset_queries) { # nolint: cyclocomp_li
     group_by_at(static_cols) %>%
     summarise_all(~ first(na.omit(.))) %>%
     ungroup()
```
The results should not be assigned to a variable, i.e., `joined <- ` should be removed. Otherwise the output dataset is not printed.

bundfussr commented on R/derive_vars_query.R:
```diff
@@ -1,5 +1,8 @@
 #' Derive Query Variables
 #'
+#' **Note:** This function exists to work with derive_vars_query, to execute
+#' the first step of derive_vars_query separately for metadata tracking
```
Either we should remove this note or we should explain what is meant by "metadata tracking".

imccay13 replied:
@paltusplintus do you have a better way to say this?

imccay13 commented:
@bundfussr thank you for your comments, I implemented them all so far other than the ones describing what we're doing here: I think @paltusplintus will come up with something better than I can so I've looped him in

bms63 commented on .gitignore:
```diff
@@ -53,4 +53,4 @@ admiral*.tgz
 
-.Rprofile
\ No newline at end of file
+.Rprofile
```
I don't think we need to re-include the Rprofile again? unless I missed something

bms63 commented on NEWS.md:
```diff
@@ -1,7 +1,10 @@
+
+
+## New Features
+- Added helper functions to derive_vars_merged() (get_flagged_records()) and derive_vars_query() (get_vars_query()) so they call the new function to perform previous functionality exactly as it was, but with more explicit metadata resulting (#2349)
```
```suggestion
- Added helper functions to `derive_vars_merged()` (`get_flagged_records()`) and `derive_vars_query()` (`get_vars_query()`) so they call the new function to perform previous functionality exactly as it was, but with more explicit metadata resulting (#2349)
```

bms63 replied:
this way they become links to the their documentations

bms63 commented on R/get_flagged_records.R:
```diff
@@ -0,0 +1,104 @@
+#' there exists at least one observation in another dataset fulfilling a certain
+#' condition.
+#'
+#' **Note:** This is a helper function for `derive_vars_merged_exist_flag` which
+#' inputs this result into `derive_vars_merged()`.
```
```suggestion
#' **Note:** This is a helper function for `derive_vars_merged_exist_flag()` which
#' inputs this result into `derive_vars_merged()`.
```

bms63 replied:
renders as a link now

bms63 commented on tests/testthat/test-get_flagged_records.R:
Can we add a test with some missing data in AVISIT - use NA_character_

bms63 replied:
I also see a filter argument for this function which is not tested.  Can you add a test that make sure the filter argument passes as well.

bms63 commented on tests/testthat/test-get_flagged_records.R:
```diff
@@ -0,0 +1,26 @@
+advs <- tibble::tribble(
+  ~USUBJID, ~PARAMCD, ~AVISIT,    ~AVAL,
+  "ST42-1", "WEIGHT", "BASELINE", 66,
+  "ST42-1", "WEIGHT", "Week 2",   68,
+  "ST42-2", "WEIGHT", "BASELINE", 88,
+  "ST42-3", "WEIGHT", "Week 2",   55,
+  "ST42-3", "WEIGHT", "Week 4",   50
+) %>% mutate(STUDYID = "ST42")
+
+# get_flagged_records ----
+## Test 1: generate existence flag ----
+test_that("get_flagged_records Test 1: generate existence flag", {
+  actual <- get_flagged_records(
+    dataset = advs,
+    new_var = VSEVALFL,
+    condition = AVISIT == "BASELINE"
+  )
```
I didn't style this, but please move the dataset into the testthat test so it doesn't spill over into other tests.

bms63 replied:
We don't always follow this very well in admiral, but try to do our best!

bms63 commented on R/get_flagged_records.R:
```diff
@@ -0,0 +1,104 @@
+#'
+#' @param dataset The additional dataset you wish to pull variables from
+#'
+#' @param new_var New variable
+#'
+#'   The specified variable is added to the input dataset.
```
I see the new variable is always coded to 1 or 0. @bundfussr should we specify this in the argument on the return value?

bundfussr replied:
I don't think we have a convention where this should be specified. For me either way is fine.

bms63 commented on R/derive_vars_query.R:
```diff
@@ -1,4 +1,7 @@
-#' Derive Query Variables
+#' Get Query Variables
+#'
+#' **Note:** This function exists to work with derive_vars_query, to execute
```
Can we get a brief summary of the purpose of the function - similar to what you did in `get_flagged_records`.  

![image](https://github.com/pharmaverse/admiral/assets/10111024/a320ca50-5cc9-4af9-b699-af9096c7d00c)

bms63 commented on tests/testthat/_snaps/derive_vars_query.md:
```diff
@@ -3,16 +3,16 @@
-      Error in `derive_vars_query()`:
+      Error in `get_vars_query()`:
       ! The source variables (values of `SRCVAR`) must be numeric or character.
       i AELLTCD is of type logical
 
-# derive_vars_query Test 6: Error is given when both TERMCHAR/TERMNUM are NA or empty
+# derive_vars_for_query Test 6: Error is given if both TERMCHAR/TERMNUM are NA/empty
 
     Code
       derive_vars_query(my_ae, query)
     Condition
-      Error in `derive_vars_query()`:
+      Error in `get_vars_query()`:
```
umm... @bundfussr this is that issue again where a function is returned with an error that the user didn't directly call.  Do you think we can update this to what you did in that big PR with all the abort stuff?

bundfussr replied:
Hm, I am not sure if this can be solved easily. We could put the `get_vars_query()` call in `derive_vars_query()` into a `tryCatch()` and issue the error with an updated `call` argument.

imccay13 replied:
added a tryCatch per this that adds a sentence about how the error is being thrown because get is called in derive - lmk what you think

bms63 reviewed (changes requested):
Thanks @imccay13 this is looking really good!!  Probably can merge in on Monday/Tuesday next week.

Please do consider some more tests with edge cases and or dirty data.  Even with 100% test coverage we always seemed to get walloped by these scenarios.

bms63 commented:
ooohhh I was just having a thought about dirty data - should a simple message be returned to the user if nothing is flagged.  I'm assuming the user would be expecting something if they are using this to derive something else...if nothing gets flagged then something might be amiss??  warning might be overkill

I guess they could also put in there own custom check...but if they go to the trouble to use this function we might want to give them something?

bundfussr commented on R/derive_vars_query.R:
```diff
@@ -35,10 +38,10 @@
 #'
 #'
-#' @return The input dataset with query variables derived.
+#' @return The processed query dataset that can be joined to the input dataset.
```
Please describe which variables are included in the output dataset.

bms63 commented:
Hi @imccay13 - we have a release planned for June 2nd.  It would be really nice to finish this PR by EOW please.  Please implement our requested changes so we can re-review.  

Also!!  many thanks for contributing!!   I know you work at GSK, but I'm not sure of your formal name.. :) Can we add you to the contributors to the Description file as well?

paltusplintus commented on NEWS.md:
```diff
@@ -1,7 +1,10 @@
+
+
+## New Features
+- Added helper functions to `derive_vars_merged()` (`get_flagged_records()`) and `derive_vars_query()` (`get_vars_query()`) so they call the new function to perform previous functionality exactly as it was, but with more explicit metadata resulting (#2349)
```
@imccay13 Perhaps change to 
```
- Added helper functions to `derive_vars_merged()` (`get_flagged_records()`) and `derive_vars_query()` (`get_vars_query()`) so that those can be called independently as per user's request. derive_vars_merged function call results are not impacted by the change.
```

bms63 commented:
@imccay13 I created the Discussion as an issue and updated the title of the PR.  This way once the PR merges into main it will automatically close the issue.  

@paltusplintus and @imccay13 can this be completed by EOW with our requested changes?  The code is fine, we just like our documentation and tests to be really clean and have pretty bows.  

Release is happening on June 2nd.

bundfussr commented on R/derive_vars_query.R:
```diff
@@ -187,14 +195,97 @@ derive_vars_query <- function(dataset, dataset_queries) { # nolint: cyclocomp_li
+#' @family der_occds
+#' @keywords der_occds
+#'
+#' @seealso [create_query_data()]]
```
```suggestion
#' @seealso [create_query_data()]
```

bundfussr commented on R/get_flagged_records.R:
```diff
@@ -0,0 +1,94 @@
+#' condition.
+#'
+#' **Note:** This is a helper function for `derive_vars_merged_exist_flag()` which
+#' inputs this result into `derive_vars_merged()`..
```
```suggestion
#' inputs this result into `derive_vars_merged()`.
```

bundfussr commented on R/get_flagged_records.R:
```diff
@@ -0,0 +1,94 @@
+#'
+#'   The condition is evaluated at the dataset (`dataset`). For
+#'   all rows where it evaluates as `TRUE` the new variable
+#'   is set to 1 in the new column
```
```suggestion
#'   is set to `1` in the new column. Otherwise, it is set to `0`.
```

bundfussr commented on R/get_flagged_records.R:
```diff
@@ -0,0 +1,94 @@
+#'
+#'   *Permitted Values*: a condition
+#'
+#' @return The output dataset is the input dataset filtered by the `filter` condition` and
```
```suggestion
#' @return The output dataset is the input dataset filtered by the `filter` condition
```

bms63 commented on tests/testthat/test-derive_vars_query.R:
```diff
@@ -156,7 +156,7 @@ test_that("derive_vars_query Test 5: Derive decides between TERMCHAR and TERMNUM
 
 ## Test 6: Error is given when both TERMCHAR/TERMNUM are NA or empty ----
-test_that("derive_vars_query Test 6: Error is given when both TERMCHAR/TERMNUM are NA or empty", {
+test_that("derive_vars_for_query Test 6: Error is given if both TERMCHAR/TERMNUM are NA/empty", {
```
I think this is a typo?  `derive_vars_for_query` should be `derive_vars_query`

bms63 commented on R/derive_vars_query.R:
```diff
@@ -1,4 +1,11 @@
+#' rows for a `derive_vars_query()` call with the relevant `SRCVAR`, `TERM_NAME_ID`
+#' and a temporary index if it is necessary
+#'
+#' **Note:** This function is the first step performed in derive_vars_query
```
```suggestion
#' **Note:** This function is the first step performed in `derive_vars_query()`
```

bms63 commented:
@imccay13 can you make these last couple of tweaks and then we can merge in!!  Also, please add yourself to description file or ping me in TEAMS and I can add you to the list

bms63 reviewed (approved):
Thanks @imccay13
